[state_dict][10/N] Let set_state_dict returns IncompatibleKeys

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112414
* #112203
* #112167

load_state_dict returns IncompatibleKeys, so set should also return the same information for the users.

Differential Revision: [D50748157](https://our.internmc.facebook.com/intern/diff/D50748157/)

fegin committed:
[state_dict][10/N] Let set_state_dict returns IncompatibleKeys

load_state_dict returns IncompatibleKeys, so set should also return the same information for the users.

Differential Revision: [D50748157](https://our.internmc.facebook.com/intern/diff/D50748157/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][10/N] Let set_state_dict returns IncompatibleKeys"


load_state_dict returns IncompatibleKeys, so set should also return the same information for the users.

Differential Revision: [D50748157](https://our.internmc.facebook.com/intern/diff/D50748157/)

[ghstack-poisoned]

wz337 commented on test/distributed/checkpoint/e2e/test_pipeline.py:
```diff
@@ -90,7 +90,7 @@ def load_with_fsdp(self, pipeline_dir: str) -> None:
         )
 
-    @skip_if_lt_x_gpu(4)
+    @skip_if_lt_x_gpu(8)
```
Just curious, do multi-gpu CI tests have 8 gpus? Does 4 not enough for this test case?

fegin replied:
oh, right, I was trying something else. Let me fix this.

wz337 reviewed (approved):
LGTM!

fegin committed:
Update on "[state_dict][10/N] Let set_state_dict returns IncompatibleKeys"


load_state_dict returns IncompatibleKeys, so set should also return the same information for the users.

Differential Revision: [D50748157](https://our.internmc.facebook.com/intern/diff/D50748157/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
